### PR TITLE
Add extensible extractor APIs

### PR DESCRIPTION
**Pull request by jaymo001**
This change adds:
1) Add SparkRowExtractor to support row-based customizable extractor
2) Refactor DataFrameBasedRowEvaluator to unify row-based customizable extractor and MVEL-based extractor

### CODE REVIEW

File: src/main/scala/com/linkedin/feathr/offline/transformation/DataFrameBasedRowEvaluator.scala [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    val (unspecifiedTypeConfigs, providedTypeConfigs) = featureTypeConfigs.partition(x => x._2.getFeatureType == FeatureTypes.UNSPECIFIED)
+    // providedTypeConfigs should take precedence over inferredFeatureTypeConfigs; inferredFeatureTypeConfigs should
+    // take precedence over unspecifiedTypeConfigs
+    val mergedTypeConfigs = unspecifiedTypeConfigs ++ inferredFeatureTypeConfigs ++ providedTypeConfigs
+    val featureDF: DataFrame = createFDSFeatureDF(inputDF, featureRefStrs, spark, transformedRdd, mergedTypeConfigs)
+    offline.FeatureDataFrame(featureDF, mergedTypeConfigs)
+  }
+
+  /**
+   * Create a FDS feature dataframe from the MVEL transformed RDD.
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/transformation/DataFrameBasedRowEvaluator.scala:108]: is FDS explained anywhere?

File: src/main/scala/com/linkedin/feathr/offline/transformation/DataFrameBasedRowEvaluator.scala [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    val featureDF = featureRefStrs
+      .zip(featureRefStrs)
+      .foldLeft(withoutDupContextFieldDF)((baseDF, namePair) => {
+        baseDF.withColumnRenamed(featureColumnNamePrefix + namePair._1, namePair._2)
+      })
+    featureDF
+  }
+}
+
+// Used to 'accumulate' the feature type for a feature while transforming the feature for each row in the dataset
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/transformation/DataFrameBasedRowEvaluator.scala:147]: we should explain 'accumulate' here.

File: src/main/scala/com/linkedin/feathr/offline/job/FeatureTransformation.scala [hunk 3]
```diff
@@ -13,7 +13,7 @@ import com.linkedin.feathr.offline.join.DataFrameKeyCombiner
 import com.linkedin.feathr.offline.source.accessor.{DataSourceAccessor, NonTimeBasedDataSourceAccessor, TimeBasedDataSourceAccessor}
 import com.linkedin.feathr.offline.swa.SlidingWindowFeatureUtils
 import com.linkedin.feathr.offline.transformation.FeatureColumnFormat.FeatureColumnFormat
-import com.linkedin.feathr.offline.transformation.{DataFrameBasedSqlEvaluator, FeatureColumnFormat, FeatureTypeAccumulator, WindowAggregationEvaluator}
+import com.linkedin.feathr.offline.transformation._
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/job/FeatureTransformation.scala:16]: our convention is: avoid wildcard imports

File: src/main/scala/com/linkedin/feathr/offline/config/FeathrConfigLoader.scala [hunk 4]
```diff
@@ -6,8 +6,8 @@ import com.fasterxml.jackson.databind._
 import com.fasterxml.jackson.databind.module.SimpleModule
 import com.fasterxml.jackson.databind.node._
 import com.jasonclawson.jackson.dataformat.hocon.HoconFactory
-import com.linkedin.feathr.common._
 import com.linkedin.feathr.common.exception.{ErrorLabel, FeathrConfigException, FeathrException}
+import com.linkedin.feathr.common.{AnchorExtractor, _}
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/config/FeathrConfigLoader.scala:10]: avoid wildcard imports

**Review by xiaoyongzhu — APPROVED** [review]

**Review by hangfei — APPROVED** [review]

**Pull request merged**